
# GenomeDiff-to-NGCircos Visualization Pipeline

**ðŸ“ˆ Interactive Visualization of Breseq Mutation Predictions with NG-Circos.js**

This repository provides a streamlined, end-to-end pipeline for converting mutation predictions from Breseqâ€™s `.gd` (GenomeDiff) files into rich, interactive circular genome visualizations using **[NG-Circos.js](https://github.com/dugongjs/NG-Circos)** (modified).

It bridges mutation annotation parsing, reference genome interpretation, and D3.js-powered visual rendering into a single, easy-to-use framework.


## Prerequisites

- [Miniconda](https://docs.conda.io/en/latest/miniconda.html) or Anaconda
- Python = 3.11
- Breseq `annotated.gd` or `output.gd` 
- Reference genome in `.gbk/.gff3/.fa` format (it should be of the same strain as in `annotated.gd` or `output.gd` )

---

## Getting Started

### 1. Clone the Repository

```bash
git clone https://github.com/PadmanabhanKann/genome_visualisation
cd genome_visualisation/master_script
```

### 2. Set Up Conda Environment

```bash
conda env create -f environment.yml -n gd-viz
conda activate gd-viz
```


## Workflow Execution

### Step 1: Process Breseq Output

```bash
python master.py \
  --ref reference.gbk \
  --out output.gd \
  --ann breseq_output/output.gd
```

**Arguments**:

- `--ref`: Path to the reference genome in GenBank format
- `--out`: `output.gd` (generates`annotated.gd` if not provided)
- `--ann`: Path to the annotated `.gd` file (Breseq output)

ðŸ“Ž To view all arguments, run:

```bash
python master.py -h
```

---

## Visualization Output

Once the script completes, open `Genome.html` in your browser to interactively explore the visualized genome.

---

## Example Output

The `example/` folder includes:

- Sample `.gd` and `.gbk` input files
- Pre-generated JSON outputs
- A demo HTML (`Genome.html`) showing how the circular genome visualization should look

This is a great starting point to understand the expected output and customize your own plots.

---

## Repository Structure

```
genome_visualisation/
â”œâ”€â”€ master_script/
â”‚   â”œâ”€â”€ amp.py
â”‚   â”œâ”€â”€ del.py
â”‚   â”œâ”€â”€ html.py
â”‚   â”œâ”€â”€ ins.py
â”‚   â”œâ”€â”€ inv.py
â”‚   â”œâ”€â”€ master.py
â”‚   â”œâ”€â”€ mob.py
â”‚   â”œâ”€â”€ snp.py
â”‚   â”œâ”€â”€ lib/
â”‚   â”‚   â”œâ”€â”€ NGCircos_.js
â”‚   â”‚   â”œâ”€â”€ d3.js
â”‚   â”‚   â”œâ”€â”€ jquery.js
â”‚   â”‚   â””â”€â”€ saveSvgAsPng.js
â”‚   â””â”€â”€ css/
â”‚       â”œâ”€â”€ bootstrap.min.css
â”‚       â”œâ”€â”€ font-awesome.min.css
â”‚       â””â”€â”€ sb-admin-2.css
â”œâ”€â”€ example/
â”‚   â”œâ”€â”€ REL606.gbk (sample reference file)
â”‚   â”œâ”€â”€ REL606.gff3(sample reference file)
â”‚   â”œâ”€â”€ annotated.gd(sample annotated.gd)
â”‚   â””â”€â”€ genome.html
â”œâ”€â”€ environment.yml
â””â”€â”€ requirements.txt

```

---

